Runtime/Scripts/DebugUtils/RGGizmos.cs

```diff
@@ -0,0 +1,396 @@
+                }
+            }
+            foreach (var lineParams in _linesFromPositionToPosition.Values)
+            {
+                Debug.DrawLine(lineParams.Item1, lineParams.Item2, lineParams.Item3);
+            }
+            
+            // Draw spheres
+            foreach (var sphereParams in _spheresAtPosition.Values)
+            {
+                if (sphereParams.Item4)
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawWireSphere(sphereParams.Item1, sphereParams.Item3);
+                }
+                else
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawSphere(sphereParams.Item1, sphereParams.Item3);
+                }
+            }
+            foreach (var sphereParams in _spheresAtEntity.Values)
+            {
+                var originInstance = RGFindUtils.Instance.FindOneByInstanceId<RGEntity>(sphereParams.Item1);
+                if (originInstance == null) continue;
+                if (sphereParams.Item4)
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawWireSphere(originInstance.transform.position, sphereParams.Item3);
+                }
+                else
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawSphere(originInstance.transform.position, sphereParams.Item3);
+                }
+            }
+            
+            // Delete any billboards that are no longer being drawn
+            var billboardsToRemove = _drawnBillboards.Where(b => !_billboardsToDraw.ContainsKey(b.Key));
+            foreach (var billboard in billboardsToRemove)
+            {
+                Object.Destroy(billboard.Value);
+                _drawnBillboards.Remove(billboard.Key, out _);
+            }
+            
+            // Now update or create any billboards that are being drawn
+            foreach (var billboardParams in _billboardsToDraw)
+            {
+                try
+                {
+                    // If the billboard does not exist, create it
+                    if (!_drawnBillboards.ContainsKey(billboardParams.Key))
+                    {
+                        var entityGameObject = RGFindUtils.Instance.FindOneByInstanceId<RGEntity>(billboardParams.Key);
+                        if (entityGameObject == null) continue;
+                        var billboardObject = Object.Instantiate(_billboardAsset, entityGameObject.transform.position,
+                            Quaternion.identity);
+                        billboardObject.transform.SetParent(entityGameObject.transform);
+                        _drawnBillboards[billboardParams.Key] = billboardObject;
+                    }
+                
+                    // Then set the parameters
+                    var billboard = _drawnBillboards[billboardParams.Key];
```

Comment:
The "`dict.ContainsKey(key)` then get/set `dict[key]`" pattern isn't thread-safe, even when working with a `ConcurrentDictionary`. Nothing prevents another thread from adding/removing the dictionary entry between when you called `ContainsKey` and when you create/update the entry.

Instead, you should use `GetOrAdd`:

```csharp
var billboard = _drawnBillboards.GetOrAdd(billboardParams.Key, (key) => {
    // This is called if the billboard didn't exist in the dictionary.
    // Initialize it and return the value.
});
```

It's important to note that your "Add" callback _may be called multiple times for the same key_. The lock that ConcurrentDictionary uses to protect against multiple concurrent calls to GetOrAdd is _only_ taken **after** the callback has run in order to protect actually placing the new object in the dictionary. So, you may still end up calling `Object.Instantiate` multiple times for the same billboard key, if `OnDrawGizmos` is run concurrently.

Do we know how `OnDrawGizmos` could be called concurrently? It seems like we should prevent that by taking a lock of some kind. If we believe that calling `OnDrawGizmos` concurrently is incorrect, there are some fairly cheap mechanisms we can use. Even a traditional lock _should_ have low-to-no contention. Trying to draw the same Gizmos concurrently seems incorrect to me, but I'm not totally clear on the usage.

Reply:
Disallowing concurrent calls to `OnDrawGizmos` would also solve other thread-safety issues with the other `ConcurrentDictionary`s in this method. IMO, it's still reasonable to use `ConcurrentDictionary`s in this type since I think it's _desirable_ to allow the `CreateN`/`DestroyN` methods to be called concurrently. I just also think it's reasonable to disallow concurrent Drawing.

Reply:
Luckily OnDrawGizmos should **not** ever be called concurrently, since it is only executed on the main thread by the Unity engine. The only concurrent part of this is that multiple threads may try to write to the dictionaries, but this particular function is only called by one thread.

So I think if I add your GetOrAdd suggestion, we should be good?

Reply:
@analogrelay alright this should be addressed properly now, I think...

Reply:
> So I think if I add your GetOrAdd suggestion, we should be good?

Yep, that should be fine. Especially since it looks like `_drawnBillboards` is only updated by `OnDrawGizmos`. I'd love it if we could have something in place to enforce that `OnDrawGizmos` is only run by a single thread, but I don't know a quick way to do that off-hand. This should be fine as-is.